server/controllers/admin/roles.js

```diff
@@ -0,0 +1,211 @@
+module.exports.create = create;
+module.exports.update = update;
+module.exports.remove = remove;
+module.exports.affectPages = affectPages;
+module.exports.affectToUser = affectToUser;
+module.exports.listForUser = listForUser;
+
+function list(req, res, next) {
+
+  const filters = new FilterParser(req.query, { tableAlias : 'r' });
+
+  let sql = `
+    SELECT BUID(r.uuid) as uuid, r.label, r.project_id 
+    FROM role  r
+  `;
+
+  filters.equals('project_id');
+  sql = filters.applyQuery(sql);
+
+  db.exec(sql, filters.parameters())
+    .then((rows) => {
+      res.json(rows);
+    }).catch(next)
+    .done();
+}
+
+
+function detail(req, res, next) {
+  const sql = `
+   SELECT BUID(uuid) as uuid, label, project_id 
+   FROM role
+   WHERE  uuid = ?  LIMIT 1
+  `;
+  const binaryUuid = db.bid(req.params.uuid);
+  db.one(sql, binaryUuid)
+    .then((rows) => {
+      res.json(rows);
+    })
+    .catch(next)
+    .done();
+}
+
+// create a new role
+function create(req, res, next) {
+  var sql = `
+    INSERT INTO  role(uuid, label, project_id)
+    VALUES(?, ?,?)
+  `;
+
+  db.exec(sql, [db.uuid(), req.body.label, req.body.project_id])
+    .then((rows) => {
+      res.json(rows);
+    })
+    .catch(next)
+    .done();
+}
+
+function update(req, res, next) {
+  const role = req.body;
+  delete role.uuid;
+  const sql = ` UPDATE role SET ? WHERE uuid = ?`;
+
+  db.exec(sql, [role, db.bid(req.params.uuid)])
+    .then((rows) => {
+      res.json(rows);
+    })
+    .catch(next)
+    .done();
+}
+
+
+function remove(req, res, next) {
+  const binaryUuid = db.bid(req.params.uuid);
+
+  const sql = `DELETE FROM role WHERE uuid = ?`;
+  db.exec(sql, binaryUuid)
+    .then((rows) => {
+      res.json(rows);
+    })
+    .catch(next)
+    .done();
+}
+
+// affect permission to a specific role
+
+function affectPages(req, res, next) {
+  const data = req.body;
+
+  const unitIds = data.unit_ids;
+  const roleUuid = db.bid(data.role_uuid);
+
+  const msg = {
+    state : 1,
+    message : 'success',
+  };
+  const deleteFromRole = `DELETE FROM role_unit WHERE role_uuid = ? `;
+  const affectPage = `
+    INSERT INTO  role_unit(uuid, unit_id, role_uuid)
+    VALUES( ?, ?, ?)`;
+
+  db.exec(deleteFromRole, roleUuid)
+    .then(() => {
+      const t = [];
+
+      if (!unitIds) {
+        return;
+      }
+      if (_.isArray(unitIds)) {
+        unitIds.forEach((id, idx) => {
+          t[idx] = db.exec(affectPage, [db.uuid(), id, roleUuid]);
+        });
+      } else {
+        t[0] = db.exec(affectPage, [db.uuid(), unitIds, roleUuid]);
+      }
+      return q.all(t);
+    })
+    .then(() => {
+      res.json(msg);
+    })
+    .catch(next)
+    .done();
+}
+
+
+// retireves affected and not affected role by a user id
+function listForUser(req, res, next) {
+  const userId = req.params.user_id;
+  const projectId = req.params.project_id;
+
+  const sql = `
+    SELECT DISTINCT q.uuid, q.label, q.project_id , q.affected
```

Comment:
If I understand your query, you want to have every role in the database, and a boolean column "affected" that will tell if it is assigned to the user or not.  This is the ideal situation for a LEFT JOIN and a GROUP BY.  Here is how it will work:
 1. The LEFT JOIN will ensure that we get every role, even if there is no corresponding value in the user_role table for that role.  (If you need practice remembering what LEFT JOINs do, [this is a good article about them](http://www.mysqltutorial.org/mysql-left-join.aspx)).
 2. The GROUP BY will make sure that we only get one value for each role, and allow us to check if the user is assigned to it.
Like this:
```sql
SELECT BUID(r.uuid) AS uuid, r.label, r.project_id, MAX(IF(ur.user_id = ?, 1, 0)) AS affected
FROM role AS r
LEFT JOIN user_role AS ur ON r.uuid = ur.role_uuid
GROUP BY role.uuid
WHERE project_id = ?;
```

Reply:
MySQL version 5.7+ may complain because GROUP BY statements should specify an aggregation value for each column.  If that happens to you, add MAX() or MIN() to each column (besides the grouping column).

Reply:
Your query is good if a role has been already assigned to a user else it returns an empty array.
also ``` GROUP BY role.uuid ```  must come after  ```WHERE project_id = ?```
So I keep mine for the moment and think of an efficient one